### PR TITLE
Refactor annotation JSON presenter tests

### PR DESCRIPTION
**Pull request by jon-betts**
For: https://github.com/hypothesis/h/issues/6769

This is to reduce the general level of crazy, but specifically:

* Stop testing code outside of this module (lots of permissions tests were duplicated here)
* Make this more isolated and not sensitive to changes in other code
* Make use of the boundary of the class for testing, instead of unpredictably mocking some stuff and not others

Also some minor shuffling and refactors of the code under test to make it more obvious

### CODE REVIEW

File: h/presenters/annotation_json.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import copy
+from copy import deepcopy
```
Comment (jon-betts) [inline h/presenters/annotation_json.py:1]: If we care about not mutating something we should probably care about not mutating it beyond the first level.

File: h/presenters/annotation_json.py [hunk 2]
```diff
@@ -9,68 +9,59 @@
...
     """Present an annotation in the JSON format returned by API requests."""
 
     def __init__(self, annotation_context, formatters=None):
         super().__init__(annotation_context)
 
-        self._formatters = []
-
-        if formatters is not None:
-            self._formatters.extend(formatters)
+        self._formatters = tuple(formatters or [])
```
Comment (jon-betts) [inline h/presenters/annotation_json.py:17]: I think this is all intended to make this immutable, which I think a tuple does quite well

File: h/presenters/annotation_json.py [hunk 3]
```diff
@@ -9,68 +9,59 @@
...
+                    "read": [self._get_read_permission()],
+                    "admin": [self.annotation.userid],
+                    "update": [self.annotation.userid],
+                    "delete": [self.annotation.userid],
+                },
+                "target": self.target,
+                "document": DocumentJSONPresenter(self.annotation.document).asdict(),
+                "links": self.links,
+            }
+        )
```
Comment (jon-betts) [inline h/presenters/annotation_json.py:45]: This is the same stuff except we start from the cloned extras and update it instead of creating a separate dict and doing that after. This made more sense to me as it means the code is laid out in the order the data is applied.

The permissions are also set here, because only one of them changes, and as far as I can tell this property is only ever used locally.

File: tests/h/presenters/annotation_json_test.py [hunk 4]
```diff
@@ -1,208 +1,119 @@
...
-            groupid="__world__",
-            shared=True,
-            target_selectors=[{"TestSelector": "foobar"}],
-            references=["referenced-id-1", "referenced-id-2"],
-            extra={"extra-1": "foo", "extra-2": "bar"},
+    def test_asdict(self, annotation, context, DocumentJSONPresenter):
+        annotation.created = datetime.datetime(2016, 2, 24, 18, 3, 25, 768)
+        annotation.updated = datetime.datetime(2016, 2, 29, 10, 24, 5, 564)
+        annotation.references = ["referenced-id-1", "referenced-id-2"]
+        annotation.extra = {"extra-1": "foo", "extra-2": "bar"}
```
Comment (jon-betts) [inline tests/h/presenters/annotation_json_test.py:18]: Set the minimum we have to locally, and then do everything else by reference.

File: tests/h/presenters/annotation_json_test.py [hunk 5]
```diff
@@ -1,208 +1,119 @@
...
+                "id": annotation.id,
+                "created": "2016-02-24T18:03:25.000768+00:00",
+                "updated": "2016-02-29T10:24:05.000564+00:00",
+                "user": annotation.userid,
+                "uri": annotation.target_uri,
+                "group": annotation.groupid,
+                "text": annotation.text,
+                "tags": annotation.tags,
+                "permissions": {
+                    "read": [annotation.userid],
```
Comment (jon-betts) [inline tests/h/presenters/annotation_json_test.py:34]: Technically we duplicate testing this below, but I thought it was better than being weird and having an `Any` here

File: tests/h/presenters/annotation_json_test.py [hunk 6]
```diff
@@ -1,208 +1,119 @@
...
-            (
-                mock.Mock(groupid="open", shared=True),
-                "world",
-                "read",
-                ["group:__world__"],
-            ),
-            (mock.Mock(userid="acct:luke"), None, "admin", ["acct:luke"]),
-            (mock.Mock(userid="acct:luke"), None, "update", ["acct:luke"]),
-            (mock.Mock(userid="acct:luke"), None, "delete", ["acct:luke"]),
-        ],
```
Comment (jon-betts) [inline tests/h/presenters/annotation_json_test.py:177]: This is the major motivator for looking at these tests. What the flipping heck is this up to?

To me this looks like a duplication of the testing done for the `AnnotationContext`. Note that in the previous version the real `AnnotationContext` was used, not a mocked one.

File: h/presenters/annotation_json.py (hunk 1 above)
Comment (seanh) [reply h/presenters/annotation_json.py:1]: Seems reasonable. Someone might have been thinking that `annotation.extra` wouldn't contain any nested mutable objects. And they might have been right as far as I know. But `deepcopy` doesn't have to rely on that assumption and I can't imagine it doing any _harm_

File: h/presenters/annotation_json.py (hunk 2 above)
Comment (seanh) [reply h/presenters/annotation_json.py:17]: The code wasn't going out of its way, as far as I can see, to make `self._formatters` immutable was it? It just happens not to mutate it / has no intention of ever mutating it?

Anyway, if it's not mutated and not intended to be mutable then a tuple seems more correct than a list since it signals to the reader that this is not going to be mutated. I think that's good usage of different builtin types

File: h/presenters/annotation_json.py [hunk 7]
```diff
@@ -9,68 +9,59 @@
...
-            "uri": self.annotation.target_uri,
-            "text": self.text,
-            "tags": self.tags,
-            "group": self.annotation.groupid,
-            "permissions": self.permissions,
-            "target": self.target,
-            "document": docpresenter.asdict(),
-            "links": self.links,
-        }
+        annotation = deepcopy(self.annotation.extra) or {}
```
Comment (seanh) [inline h/presenters/annotation_json.py:20]: I guess the `or` is because `annotation.extra` might be `None` in which case `deepcopy()` would return `None`? I'm not sure whether it's necessary because `models.Annotation.extra` is not nullable. The code was already like this though

File: h/presenters/annotation_json.py (hunk 3 above)
Comment (seanh) [reply h/presenters/annotation_json.py:45]: Yeah it seems easier to read here. Removes a local variable that was just making things harder to follow

File: tests/h/presenters/annotation_json_test.py (hunk 4 above)
Comment (seanh) [reply tests/h/presenters/annotation_json_test.py:18]: Certainly good to see `mock.Mock()` gone

File: tests/h/presenters/annotation_json_test.py (hunk 5 above)
Comment (seanh) [reply tests/h/presenters/annotation_json_test.py:34]: Yeah that's fine. A little duplication isn't the worst thing in the world

File: tests/h/presenters/annotation_json_test.py [hunk 8]
```diff
@@ -1,208 +1,116 @@
...
-        context = AnnotationContext(ann, groupfinder_service, links_service)
-        formatters = []
-
-        presenter = AnnotationJSONPresenter(context, formatters)
-        formatters.append(get_formatter({"enterprise": "synergy"}))
-        presented = presenter.asdict()
-
-        assert "enterprise" not in presented
+        for formatter in formatters:
+            formatter.format.assert_called_once_with(context)
```
Comment (seanh) [inline tests/h/presenters/annotation_json_test.py:78]: Do we also want to somehow assert that the formatters changes were applied?

File: tests/h/presenters/annotation_json_test.py (hunk 6 above)
Comment (seanh) [reply tests/h/presenters/annotation_json_test.py:177]: Yeah it's quite hard to follow. The new much smaller test is much better :+1:

File: tests/h/presenters/annotation_json_test.py [hunk 9]
```diff
@@ -1,208 +1,116 @@
...
-        context = AnnotationContext(annotation, groupfinder_service, links_service)
-        presenter = AnnotationJSONPresenter(context)
-        assert expected == presenter.permissions[action]
+        annotation.shared = shared
+        principals_allowed_by_permission.return_value = readable_by
+
+        presented = AnnotationJSONPresenter(context).asdict()
+
+        permission = permission_template.format(annotation=annotation)
+        assert presented["permissions"]["read"] == [permission]
```
Comment (seanh) [inline tests/h/presenters/annotation_json_test.py:103]: That certainly seems a lot easier to understand

File: tests/h/presenters/annotation_json_test.py [hunk 10]
```diff
@@ -218,12 +126,9 @@ def get_formatter(payload=None):
         return get_formatter
 
     @pytest.fixture
-    def policy(self, pyramid_config):
-        """Set up a fake authentication policy with a real ACL authorization policy."""
-        policy = ACLAuthorizationPolicy()
-        pyramid_config.testing_securitypolicy(None)
-        pyramid_config.set_authorization_policy(policy)
+    def DocumentJSONPresenter(self, patch):
+        return patch("h.presenters.annotation_json.DocumentJSONPresenter")
```
Comment (seanh) [inline tests/h/presenters/annotation_json_test.py:130]: Should probably be `autouse`

File: tests/h/presenters/annotation_json_test.py [hunk 11]
```diff
@@ -218,12 +126,9 @@ def get_formatter(payload=None):
...
 
     @pytest.fixture
-    def document_asdict(self, patch):
-        return patch("h.presenters.annotation_json.DocumentJSONPresenter.asdict")
+    def principals_allowed_by_permission(self, patch):
+        return patch("h.presenters.annotation_json.principals_allowed_by_permission")
```
Comment (seanh) [inline tests/h/presenters/annotation_json_test.py:134]: Should probably be `autouse` too

**Review by seanh — APPROVED** [review]
Yep, pretty confusing tests. The new version is much better :+1: 

I think in particular tests that only test code within the module under test are just a lot easier to follow. You can look at the test and the code under test side by side and have a much better chance of figuring out what the test is supposed to be testing and whether it's actually doing that. In the previous version that one long test in particular was just incomprehensible.

It looks like some of the old code may pre-date our mocking best practices about using the class boundary to be predictable about what we do and do mock and about using autospec. Again much easier to understand now.

Being less sensitive in changes to other code is just a bonus compared to the size of the other benefits

File: h/presenters/annotation_json.py (hunk 1 above)
Comment (jon-betts) [reply h/presenters/annotation_json.py:1]: Yeah, I don't think it's actually going to happen. But why not bring out the big guns?

File: h/presenters/annotation_json.py (hunk 2 above)
Comment (jon-betts) [reply h/presenters/annotation_json.py:17]: Yeah it was. There was a test dedicated to this. The original PR that brought it in also mentioned it as an explicit goal?

Honestly, I don't think it matters at all, and we can just do `self.formatter = formatters` like you normally would, but this doesn't seem like much.

File: tests/h/presenters/annotation_json_test.py (hunk 8 above)
Comment (jon-betts) [reply tests/h/presenters/annotation_json_test.py:78]: There's a separate test that asserts that the formatters results are merged in. This test's focus is exclusively on the mutated one _not_ being included.

File: tests/h/presenters/annotation_json_test.py (hunk 10 above)
Comment (jon-betts) [reply tests/h/presenters/annotation_json_test.py:130]: I've added this

File: tests/h/presenters/annotation_json_test.py (hunk 11 above)
Comment (jon-betts) [reply tests/h/presenters/annotation_json_test.py:134]: Dunzo
